classes/webservice/WebserviceSpecificManagementAttachments.php

```diff
@@ -0,0 +1,300 @@
+                    break;
+            }
+        }
+        // Need to set an object for the WebserviceOutputBuilder object in any case
+        // because schema need to get webserviceParameters of this object
+        if (isset($object)) {
+            $this->wsObject->objects['empty'] = $object;
+        }
+    }
+
+    /**
+     * Handles attachment file download
+     *
+     * @throws WebserviceException if attachment is not existing or file not available
+     *
+     * @return string[] file details
+     */
+    public function executeFileGetAndHead()
+    {
+        $a = new Attachment((int) $this->wsObject->urlSegment[2]);
+        if (!$a) {
+            throw new WebserviceException(
+                sprintf(
+                    'Attachment %d not found',
+                    $this->wsObject->urlSegment[2]
+                ),
+                [
+                    1,
+                    Response::HTTP_INTERNAL_SERVER_ERROR,
+                ]
+            );
+        }
+
+        // Physical file location
+        $file = _PS_DOWNLOAD_DIR_ . $a->file;
+        // Check if file exists
+        if (!file_exists($file)) {
+            throw new WebserviceException(
+                sprintf(
+                    'Unable to load the attachment file for attachment %d',
+                    $this->wsObject->urlSegment[2]
+                ),
+                [
+                    1,
+                    Response::HTTP_INTERNAL_SERVER_ERROR,
+                ]
+            );
+        }
+
+        // Return file details
+        return array(
+            'file' => $file,
+            'mime' => $a->mime,
+            'file_name' => $a->file_name,
+            'file_size' => $a->file_size,
+        );
+    }
+
+    /**
+     * Handles file upload
+     *
+     * Creates new attachment or replaces existing with a new file.
+     * [PUT] update existing attachment file
+     * [POST] create new attachment
+     */
+    public function executeFileAddAndEdit()
+    {
+        // Load attachment with or without id depending on method
+        $a = $this->wsObject->method == 'PUT' ? new Attachment((int) $this->wsObject->urlSegment[1]) : new Attachment();
+
+        // Check form data
+        if (isset($_FILES['file']) && is_uploaded_file($_FILES['file']['tmp_name'])) {
+            // Ensure file is within allowed size limit
+            if ($_FILES['file']['size'] > (Configuration::get('PS_ATTACHMENT_MAXIMUM_SIZE') * 1024 * 1024)) {
+                $this->wsObject->errors[] = sprintf(
+                    $this->l('The file is too large. Maximum size allowed is: %1$d kB. The file you are trying to upload is %2$d kB.'),
+                    Configuration::get('PS_ATTACHMENT_MAXIMUM_SIZE') * 1024,
+                    number_format(($_FILES['file']['size'] / 1024), 2, '.', '')
+                );
+            } else {
+                // Assign unique id
+                do {
+                    $uniqid = sha1(uniqid()); // must be a sha1
```

Comment:
Maybe check if $a->id alternatively $a->file in order to run the generation of file path.
An existing attachment should perhaps only have its file updated, not reassigned to a new file.